lib/saml_idp/service_provider.rb

```diff
@@ -21,13 +21,14 @@ def valid?
     end
 
     def valid_signature?(doc, require_signature = false, options = {})
-      if require_signature || should_validate_signature?
-        Array(certs).any? do |cert|
-          if doc.valid_signature?(fingerprint_cert(cert), options.merge(cert: cert))
-            @matching_cert = cert
-            true
-          end
+      Array(certs).any? do |cert|
```

Comment:
**question:** Since this gets called by `SamlIdp::Request#valid_signature?`, does it make sense to maybe move out the `matching_cert` instance variable assignment to that object instead? That may or may not make it easier to test but at least it seems like a more logical place (since the certificate matches the request signing cert).

If we want to push this out as-is to fix the bug and follow up with a refactor for the above I'm fine with that too, but wanted to raise the idea.

Reply:
That's a great idea, I gave it a shot, and then some of the logout specs started failing in the gem, definitely worth noting for next time we have to clean this up

Reply:
Ah well, I had hoped it might not be too bad but can't say I'm surprised. Did you add a JIRA ticket to refactor?

Reply:
LG-4875

Reply:
thanks!